bypros/proxy.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	s.following = true
+	s.stopFollow = make(chan struct{}, 1)
+
+	waitDone := sync.WaitGroup{}
+
+	apiEndpoint, err := getWsAddr(req.Target)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get ws addr: %v", err)
+	}
+
+	apiURL := fmt.Sprintf("%s/%s/%s", apiEndpoint, byzcoin.ServiceName, "StreamingRequest")
+	c, _, err := websocket.DefaultDialer.Dial(apiURL, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to dial %s: %v", apiURL, err)
+	}
+
+	streamReq := byzcoin.StreamingRequest{
+		ID: req.ScID,
+	}
+
+	buf, err := protobuf.Encode(&streamReq)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to encode streaming request: %v", err)
+	}
+
+	err = c.WriteMessage(websocket.BinaryMessage, buf)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to send streaming request: %v", err)
+	}
+
+	stopPing := make(chan struct{})
+
+	// When the stop signal is received, cancel the context and allow a new call
+	// to start following.
+	go func() {
+		<-s.stopFollow
+
+		close(stopPing)
+
+		err := c.WriteControl(websocket.CloseMessage, nil, time.Now().Add(time.Second*5))
+		if err != nil {
+			log.Warnf("failed to write close: %v", err)
+		}
+
+		c.Close()
+
+		waitDone.Wait()
+		log.LLvl1("done following")
+		s.follow <- struct{}{}
+		s.following = false
+	}()
+
+	// listen to new blocks and save them in the database
+	waitDone.Add(1)
+	go func() {
+		defer waitDone.Done()
+
+		for {
+			_, buf, err := c.ReadMessage()
+			if err != nil {
+				_, ok := err.(*websocket.CloseError)
+				if !ok {
+					// that can happen when we close the connection
+					log.Warnf("failed to read request: %v", err)
+				}
+				s.notifyStop()
+				return
+			}
+
+			streamResp := byzcoin.StreamingResponse{}
+			err = protobuf.Decode(buf, &streamResp)
+			if err != nil {
+				log.Errorf("failed to decode request: %v", err)
+				s.notifyStop()
+				return
+			}
+
+			s.followCallback(streamResp, nil)
+		}
+	}()
+
+	// keep the connection alive be regularly sending a ping
+	waitDone.Add(1)
+	go func() {
+		defer waitDone.Done()
+
+		for {
+			select {
+			case <-time.After(pingWsInterval):
+				err := c.WriteMessage(websocket.PingMessage, nil)
+				if err != nil {
+					log.Warnf("failed to ping: %v", err)
+					return
+				}
+			case <-stopPing:
+				return
+			}
+
+		}
+	}()
+
+	return &EmptyReply{}, nil
+}
+
+// notifyStop can be called multiple times to stop the current following
+// session, the one that listens to new blocks. There can be only one following
+// session at a time.
+func (s *Service) notifyStop() {
+	select {
+	case s.stopFollow <- struct{}{}:
+	default:
+	}
+}
+
+// CatchUP updates the database from a given block until the end of the chain.
+// It uses the streaming service to periodically send back the catch up state to
+// the client. This is appropriate since a catch up can be quite long.
+func (s *Service) CatchUP(req *CatchUpMsg) (chan *CatchUpResponse, chan bool, error) {
```

Comment:
As written above: why is that a call that has to be done manually by the client? For a proxy service, I would think that it tries to handle an up-to-date state all the time.

Reply:
The proxy service gives the tools for a user to keep an updated proxy: you can follow a node, and perform a catch up if wanted. Those are two different operations that one might want to do separately: I want just to catch up at a given point in time, or I want just to follow from that point in time, for example.

Making a service call doing one specific action also makes it much more manageable, as previously said in another comment. And combining two service calls is also really easy to do if wanted, it should just not be part of the service.

Reply:
Hmm - still not convinced - but it's your code. Perhaps bump it off some of the engineers in your team to see if they also think that way about a proxy.

I would at least remove the notion of `node` and replace it with a notion of `byzcoin`, which includes a `roster` that might change. As discussed on #conode lately, even if the nodes are currently very static, that is not the 'natural' behaviour of a blockchain...

Reply:
I understand you view. I was implementing the proxy with another mindset: a user chose to follow a node it trusts, and the proxy service doesn't reflect the chain, but the content of a node, like this is the case of Columbus. That's a subtle difference, and I guess that's the essence of our discussion. 

In our case, we will probably run a proxy service following our DEDIS node, which we have full control of, and that we don't plan to have occasionally off. We run it as a production service: if it is down then there is a problem that must be fixed, and we don't expect our proxy to automagically choose another node at random. We fix our production conode, and make our proxy follow again + catch up.

In the case a user really wants to "follow the chain" nothing prevents it to use a roster, do the work of checking for the best node, and rotate in case a node is down. Again I think this is not something the service should do.

What do you think @gnarula @pierluca ?

Reply:
I haven't yet had the pleasure to go through the PR, so here's my (limited) understanding and my take on it:
In principle, my understanding of this proxy is that it's designed to be deployed in a "sidecar pattern", where we're creating a separate, isolated process to index information contained in the ledger and query it later. As it points to the information contained in the ledger (as seen by the local node), we can always query the ledger itself to verify the information we obtained through the proxy.
Furthermore, as a sidecar service, it is expected that its lifecycle would be tightly coupled with that of the node it is deployed next to. If the node is down (or performing in a "degraded" mode), the sidecar service might be unable to process queries, or provide out-of-date results.
If my understanding is correct, I have no philosophical objections to the approach.

As I don't particularly want to remain ignorant and I'd love to have a more informed opinion, I've added myself to the reviewers and will go over this ASAP :-)

Reply:
From a blockchain point of view, following a specific node doesn't make sense: you should never trust one node. The whole idea of a blockchain is to trust 2/3 of the nodes to do the right thing. So if you trust just one node, you remove one of the advantages of it being part of a blockchain.

If you really just want to trust one node, then you don't need a proxy. You just put a firewall on the websocket-port of your node and only allow specific applications to access that.

Saying `a user chose to follow a node it trusts` is not something you should ever do in a blockchain. It's already bad if you have to trust a proxy, but now you have to trust the proxy _and_ the node. Having forward-links makes this a little bit less awkward than in a Proof-of-Work system, as the failing node can only stop you, but still...

Reply:
@ineiti I believe the point here is that we aren't using the proxy for trust. The blockchain is still the source of trust, but a client can use the proxy to make queries. Whatever the query returns would still need to be validated with the blockchain, but that's not the proxy's responsibility.

Furthermore, because this is thought of as a sidecar pattern, I think trusting the node and the proxy is fundamentally the same.  Both are software written by the same people, and you can expect them to be operated by the same entity on the same hardware, and to be in a "degraded" or flat-out "failing" state at the same time (due to the nature of the deployment)

Finally, I disagree that "you don't need a proxy", in so far as the proxy offers functionality (i.e. "query-ability") that the blockchain does not natively offer. 

I'm under the impression that what you're describing as an alternative is not a sidecar service, but a full blown server, connected to the roster and allowing clients to perform queries on it. If I understand correctly, this effectively would be the same as trusting "itself" as a node, which is ultimately no different to the sidecar pattern. A network outage would still cause the services to be provided in a "degraded" way and it would return outdated data.

Am I missing something here?

Reply:
Oh my - this got out of hand. Long time I didn't have a long conversation on github. I always think it's a bad medium to discuss in that kind of depth.

From a technical point of view, I still reject the PR for the following reasons:
1. depending on one node instead of the roster
2. update management by the client instead of the service
3. patching onet's 5-minute rule by re-implementing the websocket-client

I see three possibilities:
A) as the cothority-code is reaching end-of-life (C4DT will probably stop supporting it in April 2022), I close my eyes and we force-merge
B) rewrite this as a library / separate app instead of a cothority-service
C) discuss over zoom why I think the above three points need to be fixed

Reply:
So it will be A): the current code is between `welded side-car` (where the service bypros and byzcoin run on the same conode, so `ServerIdentity` is equal to the conode's ServerIdentity) and `standalone` app, where a full `Roster` would make most sense to use.

Once 3) is taken care of, I'm happy to merge.

Reply:
Regarding 3) we can't spent more time to fix/update onet. We already spent quite some time on it and we definitely need to move forward with this PR, as it is on the critical path of a project that needs to move forward.